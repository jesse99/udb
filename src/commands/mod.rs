//! Handlers for the commands users type, e.g. `bt`.
pub mod info;
pub mod misc;
pub mod tables;

pub use info::*;
pub use misc::*;
