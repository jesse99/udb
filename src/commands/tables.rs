//! Helpers for building tables using the tabled crate.
use crate::utils::Styling;
use crate::utils::uwriteln;
use std::io::Write;
use tabled::{
    builder::Builder,
    settings::{Alignment, Padding, Style, object::Columns},
};

struct TableCol {
    header: String,
    align: Alignment,
    help: String,
    fields: Vec<String>,
}

/// General table. They look like this:
/// type  offset             vaddr  paddr  file size  memory size  flags   if titles
/// ----  ------             -----  -----  ---------  -----------  -----
/// Note     580                 0      0        d6c            0    ---
/// Load    2000      55957a492000      0       1000         1000    --r
/// Load    3000      55957a493000      0          0         1000    x-r
///
/// type: the segment type                                                 if explain
/// offset: the offset into the ELF file at which the segment appears
/// ...
pub struct TableBuilder {
    cols: Vec<TableCol>,
}

impl TableBuilder {
    pub fn new() -> TableBuilder {
        TableBuilder { cols: Vec::new() }
    }

    /// Left aligned column
    pub fn add_col_l(&mut self, header: &str, help: &str) {
        debug_assert!(!self.has_col(header));
        let col = TableCol {
            header: header.to_string(),
            align: Alignment::left(),
            help: help.to_string(),
            fields: Vec::new(),
        };
        self.cols.push(col);
    }

    /// Right aligned column
    pub fn add_col_r(&mut self, header: &str, help: &str) {
        debug_assert!(!self.has_col(header));
        let col = TableCol {
            header: header.to_string(),
            align: Alignment::right(),
            help: help.to_string(),
            fields: Vec::new(),
        };
        self.cols.push(col);
    }

    /// Typically add_field! is used instead.
    pub fn add_str_field(&mut self, header: &str, value: String) {
        let col = self.find_col(header);
        if value.is_empty() {
            // For some reason empty fields screw up tabled formatting.
            col.fields.push(" ".table_field().to_string());
        } else {
            col.fields.push(value);
        }
    }

    pub fn writeln(&self, mut out: impl Write, titles: bool, explain: bool) {
        uwriteln!(out, "{}", self.table_str(titles));

        if explain {
            uwriteln!(out);
            uwriteln!(out, "{}", self.explain_str());
        }
    }

    // We need to preserve add_col ordering so we can't use a HashMap
    // but O(n) should be fine for tables.
    fn has_col(&self, header: &str) -> bool {
        self.cols.iter().any(|c| c.header == header)
    }

    fn find_col(&mut self, header: &str) -> &mut TableCol {
        self.cols.iter_mut().find(|c| c.header == header).unwrap() // programmer error to not have a col
    }

    fn table_str(&self, titles: bool) -> String {
        let height = self.cols[0].fields.len();
        let mut builder = Builder::with_capacity(height + 2, self.cols.len());
        if titles {
            let names: Vec<String> = self.cols.iter().map(|c| c.header.to_string()).collect();
            let dashes: Vec<String> = names.iter().map(|s| "-".repeat(s.len())).collect();

            let header: Vec<String> = names
                .into_iter()
                .map(|s| s.table_header().to_string())
                .collect();
            let dashes: Vec<String> = dashes
                .into_iter()
                .map(|s| s.table_sep().to_string())
                .collect();
            builder.push_record(&header);
            builder.push_record(&dashes);
        }
        for i in 0..height {
            let row: Vec<String> = self.cols.iter().map(|c| c.fields[i].clone()).collect();
            builder.push_record(&row);
        }

        let mut table = builder.build();
        for (i, col) in self.cols.iter().enumerate() {
            table.modify(Columns::one(i), col.align);
        }
        table.modify(Columns::first(), Padding::new(0, 1, 0, 0));
        table.with(Style::empty());

        table.to_string()
    }

    fn explain_str(&self) -> String {
        let explains: Vec<String> = self
            .cols
            .iter()
            .map(|c| {
                format!(
                    "{}: {}",
                    c.header.clone().explain_title(),
                    c.help.clone().explain_text()
                )
            })
            .collect();
        explains.join("\n")
    }
}

macro_rules! add_field {
    ($builder:ident, $header:literal, $value:expr) => {
        let s = format!("{}", $value);
        let s = s.table_field().to_string();
        $builder.add_str_field($header, s);
    };
    ($builder:ident, $header:literal, $format:literal, $value:expr) => {
        let s = format!($format, $value);
        let s = s.table_field().to_string();
        $builder.add_str_field($header, s);
    };
}
pub(crate) use add_field;

struct SimpleRow {
    name: String,
    value: String,
    help: String,
}

/// Table with just name and value columns. They look like this:
/// little endian        true                    these have no titles
/// 64-bit               true       
///
/// little endian: blah blah                     if explain
/// 64-bit: pointers are eight bytes
pub struct SimpleTableBuilder {
    rows: Vec<SimpleRow>,
}

impl SimpleTableBuilder {
    pub fn new() -> SimpleTableBuilder {
        SimpleTableBuilder { rows: Vec::new() }
    }

    /// Typically add_simple! is used instead.
    pub fn add_str_row(&mut self, name: &str, value: String, help: &str) {
        let row = SimpleRow {
            name: name.to_string(),
            value,
            help: help.to_string(),
        };
        self.rows.push(row);
    }

    pub fn writeln(&self, mut out: impl Write, explain: bool) {
        uwriteln!(out, "{}", self.table_str());

        if explain {
            uwriteln!(out);
            uwriteln!(out, "{}", self.explain_str());
        }
    }

    fn table_str(&self) -> String {
        let height = self.rows.len();
        let mut builder = Builder::with_capacity(height + 2, 2);
        for row in self.rows.iter() {
            let row = vec![row.name.clone(), row.value.clone()];
            builder.push_record(&row);
        }

        let mut table = builder.build();
        table.modify(Columns::one(0), Alignment::left());
        table.modify(Columns::one(1), Alignment::left());
        table.modify(Columns::first(), Padding::new(0, 1, 0, 0));
        table.with(Style::empty());

        table.to_string()
    }

    fn explain_str(&self) -> String {
        let explains: Vec<String> = self
            .rows
            .iter()
            .map(|r| {
                format!(
                    "{}: {}",
                    r.name.clone().explain_title(),
                    r.help.clone().explain_text()
                )
            })
            .collect();
        explains.join("\n")
    }
}

macro_rules! add_simple {
    ($builder:ident, $name:literal, $value:expr, $help:expr) => {
        let s = format!("{}", $value);
        let s = s.table_field().to_string();
        $builder.add_str_row($name, s, $help);
    };
    ($builder:ident, $name:literal, $format:literal, $value:expr, $help:expr) => {
        let s = format!($format, $value);
        let s = s.table_field().to_string();
        $builder.add_str_row($name, s, $help);
    };
}
pub(crate) use add_simple;
