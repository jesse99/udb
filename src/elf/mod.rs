//! Generic ELF file support. These can be both core files and executable files.
//! (The executable files are useful because they contain the bulk of the debugging
//! data needed to interpret the core files.) This module contains generic ELF suport
//! but not the support for stuff like interpreting debug symbols.
//! Quick ELF reference: https://gist.github.com/x0nu11byt3/bcb35c3de461e5fb66173071a2379779
//!
//! ELF files start with an ELF header which includes:
//! * A magic number to identify the file as an ELF file.
//! * The architecture, e.g. Linux AMD x86-64.
//! * The offset to and number of program headers.
//! * The offset to and number of section headers.
//!
//! Program headers identify segments. Segments are used by the OS to load an exe into
//! memory. A program header has type, vaddr, offset, etc. Common types are:
//! * Load - for a core file these are usually memory mapped files for the exe and DLLs, for an exe these are text (CPU instructions) and data (eg statics)
//! * Note - variety of metadata, e.g. process and signal info.
//! * TLS - thread local storage info.
//!
//! Section headers identify sections. Sections are used for static linking and don't
//! appear in core files. Section headers have name, type, vaddr, offset, size, etc.
//! There are a lot of types including for the symbol table, string table, etc.
pub mod elf_file;
pub mod elf_files;
pub mod header;
pub mod io;
pub mod notes;
pub mod sections;
pub mod segments;

pub use elf_file::*;
pub use elf_files::*;
pub use header::*;
pub use io::*;
pub use notes::*;
pub use sections::*;
pub use segments::*;
