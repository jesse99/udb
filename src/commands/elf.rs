use super::tables::{add_field, add_simple};
use crate::commands::tables::{SimpleTableBuilder, TableBuilder};
use crate::debug::SymbolIndex;
use crate::elf::{
    LoadSegment, MemoryMappedFile, ProgramHeader, SectionHeader, SectionType, StringIndex,
    VirtualAddr,
};
use crate::repl::{DebugArgs, ExplainArgs, StringsArgs};
use crate::utils;
use crate::utils::Styling;
use crate::{elf::ElfFile, elf::ElfFiles, repl::TableArgs};

fn get_file(files: &ElfFiles, exe: bool) -> &ElfFile {
    if exe {
        match &files.exe {
            Some(file) => file,
            None => {
                utils::warn("--exe was used but there is no exe: using core instead");
                files.core.as_ref().unwrap()
            }
        }
    } else {
        match &files.core {
            Some(file) => file,
            None => files.exe.as_ref().unwrap(),
        }
    }
}

pub fn info_debug(files: &ElfFiles, args: &DebugArgs) {
    let file = get_file(files, true);
    match file.get_lines() {
        Some(lines) => {
            for (i, unit) in lines.units.iter().enumerate() {
                println!("compilation unit {i}:");
                println!("   sources:");
                for source in unit.source_files.iter() {
                    match source.length {
                        Some(n) => println!("      {}/{} {} bytes", source.dir, source.file, n),
                        None => println!("      {}/{}", source.dir, source.file),
                    }
                }
                println!("   include paths:");
                for i in unit.include_paths.iter() {
                    println!("      {}", i);
                }
            }
            println!("files:");
            for f in lines.files.iter() {
                println!("   {}", f);
            }
            println!("relative addresses:");
            for (a, v) in lines.lines.iter().take(args.max_lines) {
                let f = &lines.files.get(v.file);
                println!("   0x{:x}  {}:{}:{}", a.start.0, f, v.line, v.column);
            }
            if lines.lines.len() > args.max_lines {
                println!("   ...");
            }
        }
        None => println!("Couldn't find .debug_line section"),
    }
}

pub fn info_header(files: &ElfFiles, args: &ExplainArgs) {
    let mut b = SimpleTableBuilder::new();

    let file = get_file(files, args.exe);
    add_simple!(b, "type", file.header.stype(), "type of ELF file");
    if file.reader.little_endian {
        add_simple!(
            b,
            "little endian",
            file.reader.little_endian,
            "words are being laid out in memory with the most significant byte last"
        );
    } else {
        add_simple!(
            b,
            "little endian",
            file.reader.little_endian,
            "words are being laid out out in memory with the most significant byte first"
        );
    }
    if file.reader.sixty_four_bit {
        add_simple!(
            b,
            "64-bit",
            file.reader.sixty_four_bit,
            "pointers are eight bytes"
        );
    } else {
        add_simple!(
            b,
            "64-bit",
            file.reader.sixty_four_bit,
            "pointers are four bytes"
        );
    }
    add_simple!(
        b,
        "osabi",
        file.header.abi(),
        "the OS the binary was compiled for"
    );
    add_simple!(b, "abiversion", file.header.abiversion, "zero for Linux");
    add_simple!(b, "machine", file.header.machine(), "CPU architecture");
    add_simple!(b, "flags", file.header.flags, "Linux has no defined flags");
    add_simple!(
        b,
        "ph_offset",
        file.header.ph_offset,
        "offset in the ELF file to the Program Header table"
    );
    add_simple!(
        b,
        "num_ph_entries",
        file.header.num_ph_entries,
        "number of entries in the Program Header table"
    );
    add_simple!(
        b,
        "section_offset",
        file.header.section_offset,
        "offset in the ELF file to the section header table"
    );
    add_simple!(
        b,
        "num_section_entries",
        file.header.num_section_entries,
        "number of entries in the section header table"
    );
    add_simple!(
        b,
        "string_table_index",
        file.header.string_table_index,
        "section index containing the string table"
    );
    b.println(args.explain);
}

pub fn info_loads(files: &ElfFiles, args: &TableArgs) {
    pub fn find_file(
        files: &Option<Vec<MemoryMappedFile>>,
        vaddr: VirtualAddr,
    ) -> Option<&MemoryMappedFile> {
        if let Some(maps) = files {
            maps.iter().find(|m| m.vbytes.contains(vaddr))
        } else {
            None
        }
    }

    pub fn is_stack(file: &ElfFile, segment: &LoadSegment) -> bool {
        if let Some(status) = file.find_prstatus() {
            let bottom = status.get_frame_stack_bottom();
            segment.vbytes.contains(bottom)
        } else {
            false
        }
    }

    let mut builder = TableBuilder::new();
    builder.add_col_l("vaddr", "the virtual address the segment starts at");
    builder.add_col_r("vbytes", "the size of the segment in memory");
    builder.add_col_r("flags", "executable, writeable, and/or readable");
    builder.add_col_r(
        "offset",
        "the offset into the ELF file at which the segment appears",
    );
    builder.add_col_r("obytes", "the size of the segment in the core");
    builder.add_col_l(
        "note",
        "path to memory mapped file or how the segment is used",
    );

    let file = get_file(files, args.exe);
    let files = file.get_memory_mapped_files();
    for segment in file.loads.iter() {
        let mut note = String::new();
        if let Some(file) = find_file(files, segment.vbytes.start) {
            note.push_str(&format!("{} ", file.file_name));
        } else if is_stack(file, segment) {
            note.push_str("[stack] ");
        } else if !segment.executable() && segment.writeable() && segment.readable() {
            note.push_str("[data] ");
        } else if !segment.executable() && !segment.writeable() && segment.readable() {
            // TODO may also want to check that the first bytes are '.ELF'
            note.push_str("[text] ");
        }

        add_field!(builder, "vaddr", "{:x}", segment.vbytes.start.0);
        add_field!(builder, "vbytes", "{:x}", segment.vbytes.size);
        add_field!(builder, "flags", segment.flags());
        add_field!(builder, "offset", "{:x}", segment.obytes.start.0);
        add_field!(builder, "obytes", "{:x}", segment.obytes.size);
        add_field!(builder, "note", note);
    }

    builder.println(args.titles, args.explain);
}

pub fn info_notes(files: &ElfFiles, args: &TableArgs) {
    let mut builder = TableBuilder::new();
    builder.add_col_l("name", "note namespace");
    builder.add_col_l("type", "the type of the note");
    builder.add_col_r("offset", "offset into the ELF file (hex)");
    builder.add_col_r("size", "size of the note");

    let file = get_file(files, args.exe);
    for note in file.notes.iter() {
        add_field!(builder, "name", note.name);
        add_field!(builder, "type", "{:?}", note.ntype);
        add_field!(builder, "offset", "{:x}", note.contents.start.0);
        add_field!(builder, "size", note.contents.size);
    }

    builder.println(args.titles, args.explain);
}

pub fn info_relocations(files: &ElfFiles, args: &TableArgs) {
    // TODO probably should use an arg w/o --exe
    let mut builder = TableBuilder::new();
    builder.add_col_r("symbol", "name of the symbol to relocate");
    builder.add_col_r("index", "index of the symbol to relocate"); // TODO get rid of this
    builder.add_col_r("string", "index of the symbol string"); // TODO get rid of this
    builder.add_col_r("dynamic", "true if the symbol is from a shared library");
    builder.add_col_r("offset", "vaddr for exe or shared object");
    builder.add_col_l("type", "how to apply the relocation (arch specific)");
    builder.add_col_r("addend", "optional constant applied during relocation");

    let file = get_file(files, true);
    let symbols = file.find_symbols();
    let dynamic_symbols = file.find_dynamic_symbols();

    let rels = files.find_relocations();
    for r in rels.iter() {
        // TODO names aren't great for static relocation entries. They do match what
        // `readelf --syms` reports but they are sucky names. For example,
        // `readelf --relocs` will report "printf@GLIBC_2.2.5" but we say
        // "deregister_tm_clones".
        let name = if r.dynamic {
            match &dynamic_symbols {
                Some(t) => {
                    let e = t.entries.get(r.symbol as usize);
                    e.map(|ue| file.find_string(t.section.link, ue.name))
                }
                None => None,
            }
        } else {
            match &symbols {
                Some(t) => {
                    let e = t.entries.get(r.symbol as usize);
                    e.map(|ue| file.find_string(t.section.link, ue.name))
                }
                None => None,
            }
        }
        .flatten()
        .unwrap_or(format!("index {}", r.symbol));

        let string = if r.dynamic {
            match &dynamic_symbols {
                Some(t) => {
                    let e = t.entries.get(r.symbol as usize);
                    e.map(|ue| ue.name)
                }
                None => None,
            }
        } else {
            match &symbols {
                Some(t) => {
                    let e = t.entries.get(r.symbol as usize);
                    e.map(|ue| ue.name)
                }
                None => None,
            }
        }
        .unwrap_or(StringIndex(0));

        let addend = match r.addend {
            Some(a) => format!("{}", a),
            None => "none".to_string(),
        };
        add_field!(builder, "symbol", name);
        add_field!(builder, "dynamic", r.dynamic);
        add_field!(builder, "index", r.symbol);
        add_field!(builder, "string", string.0);
        add_field!(builder, "offset", "{:x}", r.offset);
        add_field!(builder, "type", "{:?}", r.rtype);
        add_field!(builder, "addend", addend);
    }

    builder.println(args.titles, args.explain);
}

pub fn info_sections(files: &ElfFiles, args: &TableArgs) {
    let file = get_file(files, args.exe);
    let sections = file.get_sections();

    let mut builder = TableBuilder::new();
    builder.add_col_r("index", "index into sections.");
    builder.add_col_l("name", "index into the string table. Zero means no name.");
    builder.add_col_l("type", "type of the section.");
    builder.add_col_r("vaddr", "virtual address at execution.");
    builder.add_col_r(
        "offset",
        "offset into the ELF file for the start of the section.",
    );
    builder.add_col_r("size", "section size in bytes.");
    builder.add_col_r("entry_size", "set if the section holds a table of entries.");
    builder.add_col_r("align", "section alignment.");
    builder.add_col_r(
        "link",
        "link to another section with related information, usually a string or symbol table.",
    );
    builder.add_col_r("info", "additional section info");
    builder.add_col_l("flags", "write, alloc, and/or exec.");

    // Would be kind of nice to sort these by name but they are referenced sometimes
    // by index...
    for (i, section) in sections.iter().enumerate() {
        add_field!(builder, "index", i); // sections are often referenced by index so this is handy
        match file.find_default_string(section.name) {
            Some(n) => {
                add_field!(builder, "name", n);
            }
            None => {
                add_field!(builder, "name", section.name.0);
            }
        };
        add_field!(builder, "type", "{:?}", section.stype);
        add_field!(builder, "flags", SectionHeader::flags(section.flags));
        add_field!(builder, "vaddr", "{:x}", section.vbytes.start.0);
        add_field!(builder, "offset", "{:x}", section.obytes.start.0);
        add_field!(builder, "size", section.vbytes.size);
        add_field!(builder, "entry_size", section.entry_size);
        add_field!(builder, "align", section.align);
        add_field!(builder, "link", section.link.0);
        add_field!(builder, "info", section.info);
    }

    builder.println(args.titles, args.explain);
}

pub fn info_segments(files: &ElfFiles, args: &TableArgs) {
    let file = get_file(files, args.exe);
    let segments = ElfFile::find_segments(file.reader, &file.header);

    let mut builder = TableBuilder::new();
    builder.add_col_l("type", "the segment type");
    builder.add_col_r(
        "offset",
        "the offset into the ELF file at which the segment appears",
    );
    if file.is_core() {
        builder.add_col_r("vaddr", "the virtual address the segment starts at");
    }
    builder.add_col_r("file size", "the size of the segment on disk");
    if file.is_core() {
        builder.add_col_r("memory size", "the size of the segment in memory");
    }
    builder.add_col_r("flags", "executable, writeable, and/or readable");

    for segment in segments.iter() {
        add_field!(builder, "type", "{:?}", segment.stype);
        add_field!(builder, "offset", "{:x}", segment.offset);
        if file.is_core() {
            add_field!(builder, "vaddr", "{:x}", segment.vaddr);
        }
        add_field!(builder, "file size", "{:x}", segment.file_size);
        if file.is_core() {
            add_field!(builder, "memory size", "{:x}", segment.mem_size);
        }
        add_field!(builder, "flags", "{}", ProgramHeader::flags(segment.flags));
    }

    builder.println(args.titles, args.explain);
    if args.explain {
        println!();
        println!("Numeric fields are all in hex. Usually it's more informative to use");
        println!("other commands like `info loads` or `info mapped`.");
    }
}

pub fn info_strings(files: &ElfFiles, args: &StringsArgs) {
    let file = get_file(files, true);
    let num_sections = file.get_sections().len();

    let mut found = false;
    for index in 0..num_sections {
        if args.index.unwrap_or(index) == index {
            let section = &file.get_sections()[index];
            if section.stype == SectionType::StringTable {
                if found {
                    println!();
                }
                println!("section {index}");
                let strings = file.find_strings(section, args.max_results);
                for (i, s) in strings.iter().enumerate() {
                    println!("{i}: {s}");
                }
                if strings.len() == args.max_results {
                    println!("...");
                }
                found = true;
            }
        }
    }
}

pub fn info_symbols(files: &ElfFiles, args: &TableArgs) {
    let mut builder = TableBuilder::new();
    builder.add_col_r("index", "symbol index");
    builder.add_col_l("name", "the symbol name");
    builder.add_col_l("type", "the symbol type");
    builder.add_col_l("dynamic", "true if the symbol is from a shared lib");
    builder.add_col_r("value", "address, absolute value, etc (in hex)");
    builder.add_col_r("size", "size of the value, 0 for unknown or undefined");
    builder.add_col_l("binding", "linkage visibility and behavior");
    builder.add_col_l(
        "visibility",
        "whether the symbol is visible outside its object file",
    );
    builder.add_col_l(
        "related",
        "indicates a related section or marks the entry as an absolute value",
    );

    // TODO double check that function pointers are legit
    // TODO sort rows? provide some sort of generic table sort support?
    // TODO maybe also filtering options, eg max-results and filter by col value
    //      two options for filter by col? or something like --filter="type=Func"?
    //      maybe also a complement option
    let file = get_file(files, args.exe);
    let tables = [file.find_dynamic_symbols(), file.find_symbols()];
    let tables = tables.iter().flatten().collect::<Vec<_>>();
    for table in tables.iter() {
        println!("using section {}", table.section.link.0);
        for (i, e) in table.entries.iter().enumerate() {
            // TODO function names can be really long (especially with name mangling)
            // readelf puts a pretty small cap on these, maybe we should default to the same
            let name = file
                .find_string(table.section.link, e.name)
                .unwrap_or("unknown".to_string());
            let name = format!("{} ({})", name, e.name.0);
            add_field!(builder, "index", i);
            add_field!(builder, "name", name);
            add_field!(builder, "dynamic", table.dynamic);
            add_field!(builder, "value", "{:x}", e.value);
            add_field!(builder, "size", e.size);
            add_field!(builder, "type", "{:?}", e.stype);
            add_field!(builder, "binding", "{:?}", e.binding);
            add_field!(builder, "visibility", "{:?}", e.visibility);
            add_field!(builder, "related", index_to_str(file, e.index));
        }
    }

    builder.println(args.titles, args.explain);
}

fn index_to_str(file: &ElfFile, index: SymbolIndex) -> String {
    match index {
        SymbolIndex::Abs => "Value".to_string(),
        SymbolIndex::Common => "Common".to_string(),
        SymbolIndex::Index(i) => file
            .find_section_name(i)
            .unwrap_or("bad section index".to_string()),
        SymbolIndex::Undef => "".to_string(),
        SymbolIndex::XIndex => "not implemented".to_string(), // TODO
    }
}
