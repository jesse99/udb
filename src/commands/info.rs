use super::tables::{add_field, add_simple};
use crate::commands::tables::{SimpleTableBuilder, TableBuilder};
use crate::elf::VirtualAddr;
use crate::repl::{ExplainArgs, LineArgs, RegistersArgs};
use crate::utils;
use crate::utils::Styling;
use crate::{elf::ElfFile, elf::ElfFiles, repl::TableArgs};
use std::cmp::Ordering;

fn get_file(files: &ElfFiles, exe: bool) -> &ElfFile {
    if exe {
        match &files.exe {
            Some(file) => file,
            None => {
                utils::warn("--exe was used but there is no exe: using core instead");
                files.core.as_ref().unwrap()
            }
        }
    } else {
        match &files.core {
            Some(file) => file,
            None => files.exe.as_ref().unwrap(),
        }
    }
}

pub fn info_line(files: &ElfFiles, args: &LineArgs) {
    match files.find_line(VirtualAddr(args.addr)) {
        Ok((file, line, col)) => println!("{file}:{line}:{col}"),
        Err(e) => println!("{e}"),
    }
}

pub fn info_mapped(files: &ElfFiles, args: &TableArgs) {
    let file = get_file(files, args.exe);
    if let Some(files) = file.get_memory_mapped_files() {
        let mut builder = TableBuilder::new();
        builder.add_col_l(
            "start",
            "the virtual address for the first byte the file is mapped into",
        );
        builder.add_col_r(
            "end",
            "the virtual address after the last byte the file is mapped into",
        );
        builder.add_col_r("size", "the size of the file in memory (decimal)");
        builder.add_col_l("file name", "path to the file");

        for file in files {
            add_field!(builder, "start", "{:x}", file.vbytes.start.0);
            add_field!(builder, "end", "{:x}", file.vbytes.end().0);
            add_field!(builder, "size", file.vbytes.size);
            add_field!(builder, "file name", file.file_name);
        }

        builder.println(args.titles, args.explain);
    } else {
        println!("No memory mapped files found.");
    }
}

pub fn info_process(files: &ElfFiles, args: &ExplainArgs) {
    let file = get_file(files, args.exe);
    if let Some(status) = file.find_prstatus() {
        let mut b = SimpleTableBuilder::new();

        add_simple!(
            b,
            "pid",
            status.pid,
            "the process id for the exe that produced the file"
        );
        add_simple!(
            b,
            "file",
            file.path.display(),
            "path to the ELF file that was loaded"
        );

        b.println(args.explain);
    } else {
        println!("No prstatus found");
    }
}

pub fn info_registers(files: &ElfFiles, args: &RegistersArgs) {
    // These come out in a really annoying order so we'll sort them.
    let file = get_file(files, args.exe);
    if let Some(status) = file.find_prstatus() {
        let mut tuples: Vec<(&'static str, u64)> = status
            .registers
            .iter()
            .enumerate()
            .filter_map(|(i, value)| {
                if args.all || !status.is_rare_register(i) {
                    let name = status.register_name(i);
                    if name != "?" {
                        return Some((name, *value));
                    }
                }
                None
            })
            .collect();

        tuples.sort_by(|lhs, rhs| {
            let lhs_num = lhs.0[1..].parse::<i32>();
            let rhs_num = rhs.0[1..].parse::<i32>();
            if let Ok(n1) = lhs_num
                && let Ok(n2) = rhs_num
            {
                // numeric registers are sorted by value, eg r9 before r11
                n1.cmp(&n2)
            } else if lhs_num.is_ok() {
                // alpha registers appear before numeric, eg rbp before r10
                Ordering::Greater
            } else if rhs_num.is_ok() {
                // alpha registers appear before numeric, eg rbp before r10
                Ordering::Less
            } else {
                // alpha registers are sorted as is, eg rbp before rip
                lhs.cmp(rhs)
            }
        });

        let mut builder = TableBuilder::new();
        builder.add_col_l("name", "the register name");
        builder.add_col_r("hex", "the register value in hex");
        builder.add_col_r("decimal", "the register value in decimal");

        for (name, value) in tuples.iter() {
            add_field!(builder, "name", name);
            add_field!(builder, "hex", "{:x}", value);
            add_field!(builder, "decimal", value);
        }

        builder.println(args.titles, args.explain);

        if args.explain {
            // TODO really these are x86 only
            utils::explain(
                "rip",
                "points to the instruction pointer currently being executed",
            );
            utils::explain(
                "rsp",
                "points to the bottom of the stack, local variables appear after this",
            );
            utils::explain(
                "rbp",
                "points to the top of the stack (depending on compiler options)",
            );
        }
    } else {
        println!("No prstatus found");
    }
}

pub fn info_signals(files: &ElfFiles, args: &TableArgs) {
    let file = get_file(files, args.exe);
    let maybe_status = file.find_prstatus();
    let maybe_signal = file.find_signal_info();

    if let Some(status) = &maybe_status {
        println!("{}", status.signal()); // this one does a nice job formatting signal and code
    } else {
        utils::warn("Couldn't find prstatus note");
    }

    if let Some(info) = &maybe_signal {
        let mut b = SimpleTableBuilder::new();
        match &info.details {
            crate::elf::SignalDetails::Fault(details) => {
                add_simple!(
                    b,
                    "fault addr",
                    "0x{:x}",
                    details.fault_addr,
                    "the address that caused the core"
                );
            }
            crate::elf::SignalDetails::Kill(details) => {
                add_simple!(
                    b,
                    "sender pid",
                    details.sender_pid,
                    "the pid of the process that sent the signal"
                );
                add_simple!(
                    b,
                    "sender uid",
                    details.sender_uid,
                    "the uid of the process that sent the signal"
                );
            }
            crate::elf::SignalDetails::Posix(details) => {
                add_simple!(
                    b,
                    "sender pid",
                    details.sender_pid,
                    "the pid of the process that sent the signal"
                );
                add_simple!(
                    b,
                    "sender uid",
                    details.sender_uid,
                    "the uid of the process that sent the signal"
                );
            }
            crate::elf::SignalDetails::Child(details) => {
                add_simple!(
                    b,
                    "child_pid",
                    details.child_pid,
                    "pid of the child process"
                );
                add_simple!(
                    b,
                    "child_uid",
                    details.child_uid,
                    "uid of the child process"
                );
                add_simple!(
                    b,
                    "exit_code",
                    details.exit_code,
                    "exit code of the child process"
                );
            }
            _ => (),
        }
        b.println(args.explain);
    } else {
        utils::warn("Couldn't find signal note");
    }
}
