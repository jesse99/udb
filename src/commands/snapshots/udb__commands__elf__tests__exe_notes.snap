---
source: src/commands/elf.rs
expression: s
---
name  type            offset  size 
----  ----            ------  ---- 
GNU   Gnu(PropType0)     348    32 
GNU   Gnu(BuildId)       378    20 
GNU   Gnu(AbiTag)        39c    16 

name  type            offset  size 
----  ----            ------  ---- 
GNU   Gnu(PropType0)     348    32 
GNU   Gnu(BuildId)       378    20 
GNU   Gnu(AbiTag)        39c    16
